### ISSUE TITLE
Cross-Modal Contrastive Learning for Text-to-Image Generation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#124]
## TL;DR

- Leverage global + local contrastive learning objective for Text-to-Image (T2I) task.
- BERT + CLIP-like Discriminator + Attn-self-modulated Generator

## Motivation

- Apply insights from contrastive learning (mutual information maximization) to T2I task

## Previous Works

- StackGAN: 2-stage conditional GAN
- AttnGAN: employ attention mechanism
- DM-GAN: adaptive refinement of generated images

## Method

![image](https://user-images.githubusercontent.com/46308407/143992200-070da77e-8106-437d-b4e9-a0a3684478b1.png)

#### Generator
- (real, fake) corresponding to the same caption
- (fake, caption)
- (fake_local, caption_word)

#### Discriminator
- (real img, caption)
- (real_local, caption_word)

![image](https://user-images.githubusercontent.com/46308407/143992419-f1c36963-d360-48c1-a48e-dd041ccf8fff.png)

### (2) Attentional Self-Modulation Generator

- Given local image feature, refine the language context based on attention coefficients

![image](https://user-images.githubusercontent.com/46308407/143992488-9fa41305-0a44-470f-8c49-b63d7b5b91b1.png)

- Obtain modulation parameters \gamma, \beta from concatenation of the global feature and context feature

![image](https://user-images.githubusercontent.com/46308407/143992538-4ae70e47-4571-4fc5-8e38-28cc81e1fbbd.png)

![image](https://user-images.githubusercontent.com/46308407/143992575-868095b8-147a-4c48-b8e9-332a5a3c76ea.png)

### (3) Contrastive Discriminator

- Role of discriminator: 1) critique in GAN 2) Context encoder (for contrastive learning)
- The discriminator only considers **real** samples for contrastive loss, since incorporating fake samples can hurt the learning dynamics of the discriminator as *context encoder*
- Conditional adversarial framework is completed with (similarity of sample and text query) + (realism logit)

![image](https://user-images.githubusercontent.com/46308407/143992800-6cf306a4-1882-4de7-8455-e92e1b19481f.png)

## Experiments

![image](https://user-images.githubusercontent.com/46308407/143992834-bbf979a7-3544-4ec0-86ef-338b29576c78.png)
![image](https://user-images.githubusercontent.com/46308407/143992849-344912c1-93d9-4546-b69c-1b6d4677e714.png)
![image](https://user-images.githubusercontent.com/46308407/143992863-9a95aa41-717f-42a9-8f32-437023975fd1.png)
![image](https://user-images.githubusercontent.com/46308407/143992876-73db5780-687a-400e-abab-ce69dcc379bc.png)
![image](https://user-images.githubusercontent.com/46308407/143992897-d9027f52-d084-47cf-a4cb-0d2ae47cfe33.png)
![image](https://user-images.githubusercontent.com/46308407/143992928-bce45df8-3d3e-46a9-be0c-3134a529401d.png)

## Comments

- Much efforts have been made to incorporate insights from self-supervised learning literature (e.g., contrastive learning) to generative modeling. This is, in my personal opinion, one of the most successful marriage between GAN and contrastive learning.
- Text-to-image task has inherent ambiguity as even for humans, picturing the exact scene given a text query is under-defined. Also, for normal discriminators, the conditional variable is too fine grained (as opposed to, say, 100 pre-defined class labels) to accurately model the conditional probability distribution. Hence, contrastive learning comes to great rescue.
- The boundary between traditional modalities (e.g., image, text, video...) is increasingly dimming, and arguably the Transformer is in the center of the revolution. Although the authors employed pretrained BERT model, which is the transformer encoder, I wonder what will happen when we model the entire framework with Transformers, instead of using traditional residual building blocks.
